### PR TITLE
feat(cli): use Config for Etherscan and RPCs

### LINKED ISSUE #3824: cast command doesn't resolve config `[etherscan]`
### Component

Cast

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

cast 0.2.0 (250cc85 2022-12-02T19:51:40.34571Z)

### What command(s) is the bug in?

cast etherscan-source --chain mainnet  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2

### Operating System

macOS (Apple Silicon)

### Describe the bug

I don't know if I miss something to config or what else.

## global config

`$HOME/.foundry.toml`

```toml
[rpc_endpoints]
mainnet  = "xxxx"

[etherscan]
mainnet = { key = "xxxx", chain = "mainnet" }
```

## run command with global config

```sh
cast etherscan-source --chain mainnet -d weth  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2

Error:
No Etherscan API Key is set. Consider using the ETHERSCAN_API_KEY env var, or setting the -e CLI argument or etherscan-api-key in foundry.toml
```

### LINKED ISSUE #4057: Multiples API Keys (BscScan, Etherscan, etc..)
### Component

Cast

### Describe the feature you would like
Hello there, 
For now using `cast` will use the `foundry.toml` which is perfect! The issue is when we want to have multiples keys on the same `foundry.toml` like here ↓
image

This will not work... Actually, only the `etherscan_api_key` seems to be chosen even when using the `-c BSC` to change the chain.
 cmd used : 
```bash
cast interface 0xc0e8D30D2ead2C324b3f1A8386992Ba1Be534CbF  -c bsc
```

### Additional context

The intented behavior should be to use the key associated to `bsc` or any others network to go faster.

### LINKED ISSUE #4401: Using a --fork-url alias and --resume together results in "Invalid provider url: <alias name>"
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (624797e 2023-02-20T00:11:38.504907Z)

### What command(s) is the bug in?

forge script

### Operating System

macOS (Apple Silicon)

### Describe the bug

It is currently not possible to continue a forge script execution (for instance if verification failed) with `--resume` while using an alias name from `foundry.toml` as the argument to `--fork-url`. The error will be

```
Invalid provider url: <alias name>

If you were trying to resume or verify a multi chain deployment, add `--multi` to your command invocation.
```

Adding `--multi` (wich should not be necessary) yields a different error: "Multi-chain deployment not found."

### LINKED ISSUE #4474: cast block-number ignores FOUNDRY_PROFILE
### Component

Cast

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (f96e0ba 2023-03-03T01:47:26.815053Z)

### What command(s) is the bug in?

cast block-number

### Operating System

macOS (Apple Silicon)

### Describe the bug

Running `FOUNDRY_PROFILE=lattice-testnet cast block-number` prints the block number of my local anvil chain, not of the RPC configured for profile `lattice-testnet` in my `foundry.toml`.

foundry.toml:
``` 
[profile.lattice-testnet]
eth_rpc_url = "https://follower.testnet-chain.linfra.xyz"
```

`cast block-number --rpc-url https://follower.testnet-chain.linfra.xyz` returns the correct result.

### LINKED ISSUE #4484: Foundry does not parse [etherscan] section of foundry.toml
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (28b2ae6 2023-03-06T00:04:12.952149891Z)

### What command(s) is the bug in?

forge verify-contract

### Operating System

Linux

### Describe the bug

1) With these settings:
```
[etherscan]
polygon = { key = "${POLYGON_ETHERSCAN_API_KEY}", url = "https://api.polygonscan.com/api" }
```
Using `forge verify-contract --chain polygon <Address> <Name>` gives:

`ETHERSCAN_API_KEY must be set`

Though key is set in the foundry.toml

2) With these settings:
```
[etherscan]
boba = { key = "${BOBA_ETHERSCAN_API_KEY}", url = "https://api.bobascan.com/api" }
```

Using `forge verify-contract --chain boba<Address> <Name>` gives:

`[ETHERSCAN_API_KEY must be set](error: invalid value 'boba' for '--chain <CHAIN>': Expected known chain or integer, found: boba)`

Though the settings for **boba** are set

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Closes #4484
Closes #4474
Closes #4401
Closes #4057
Closes #3824

## Solution

Refactor `EthereumOpts` into `EtherscanOpts` and `RpcOpts`, which implement figment traits, for usage in Cast and Forge subcommands since `rpc_url` and the rest were always separate and had inconsistent flags/envs etc...

## BREAKING

`--remappings` and `--contracts`'s short flags are now `-R` and `-C` (uppercase instead of lowercase) due to conflicts with `--rpc-url` and `--chain` respectively, while `--confirmations`'s short flag in Forge script has been removed

This is done to have consistent flags throughout all CLIs since for most of cast  `-r` is `rpc-url` while some of forge has `-r` as `--remappings`

### CODE REVIEW

File: cli/src/cmd/cast/storage.rs [hunk 1]
```diff
@@ -25,55 +29,38 @@ const MIN_SOLC: Version = Version::new(0, 6, 5);
...
 
 impl StorageArgs {
     pub async fn run(self) -> Result<()> {
-        let Self { address, block, build, rpc_url, slot, chain, etherscan_api_key } = self;
+        let Self { address, slot, block, rpc, etherscan, build } = self;
 
-        let rpc_url = try_consume_config_rpc_url(rpc_url)?;
-        let provider = try_get_http_provider(rpc_url)?;
+        let figment = Figment::from(Config::figment()).merge(etherscan).merge(rpc);
+        let config = Config::from_provider(figment);
```
Comment (mattsse) [inline cli/src/cmd/cast/storage.rs:61]: should we also merge the build args here?

worth using the figment macros+Provider impl for this type?

**Review by mattsse — COMMENTED** [review]
incredible!
you love to see it.

tysm!!!

this looks great, as usual.
smol q re StorageArgs figment.

**Comment by sakulstra** [comment]
@DaniPopes don't see it mentioned above, but checking the code this might also resolve #4401 right?

**Comment by sakulstra** [comment]
Sorry one other thing, as i saw there's now a breaking change with remappings :sweat_smile: 
I think there's one other inconsistency as current remappings in `foundry.toml` accepts an array where `--remappings` only supports a single string. Not sure if there is an issue for that.

**Comment by DaniPopes** [comment]
I don't think so, the only thing I modified here is the short flag from 'r' to 'R'; to pass multiple remappings you have to use the flag multiple times like so: `forge build -R a=b -R c=d`, and I think it was always this way

**Review by mattsse — APPROVED** [review]
🙏

**Pull request merged**